### ISSUE TITLE
ApplicationInsights context set to undefined

### ISSUE THREAD

**Issue by atbe-crowe** [MarkPieszak/angular-application-insights#47]
Hi,

---
**Version information**

- `@markpieszak/ng-application-insights`:  4.0.2
- Angular: 5.2.9
- rxjs: 5.5.7
---

I have an `AnalyticsService` I use to encapsulate the `AppInsightsService` from this library. My service looks something like this:

```typescript
import { Injectable} from "@angular/core";
import { AppInsightsService } from "@markpieszak/ng-application-insights";

@Injectable()
export class AnalyticsService {
	constructor(
		private appInsightsService: AppInsightsService) {}

	public init(): void {
		this.setInstrumentationKey();
		this.appInsightsService.init();

		this.addAnalyticsFiltering();
	}

	private setInstrumentationKey(): void {
                ...
	}

	private addAnalyticsFiltering(): void {
		this.appInsightsService.queue.push(() => {
			console.log("Adding to queue");
			console.log(this.appInsightsService.queue);
			this.appInsightsService.context.addTelemetryInitializer((envelope: Microsoft.ApplicationInsights.IEnvelope) => {
				console.log("Adding to context");
				console.log(this.appInsightsService.context);

				envelope.data.baseData.url = envelope.data.baseData.url.split("/")
					.map(part => part.split("?")[0])
					.map(part => part.split("#")[0])
					.filter(part => !part.match(/^\d+$|^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/))
					.join("/");

				return true;
			});
		});
	}
}
```

My `AnalyticsModule` looks like this:

```typescript
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { AnalyticsService } from "./analytics.service";
import { ApplicationInsightsModule, AppInsightsService } from "@markpieszak/ng-application-insights";
import { environment } from "../../../../environments/environment";

@NgModule({
	imports: [
		CommonModule,
		ApplicationInsightsModule.forRoot({
			instrumentationKeySetlater: true
		})
	],
	providers : [
		AnalyticsService
	],
	declarations: []
})
export class AnalyticsModule {
}
```

As you can see, I'm setting my instrumentation key later and have an `init` function that's called by my `AppComponent` (so that the service can start tracking all pageviews/ajax calls/exceptions right away).

Now the issue I'm having is in the `addAnalyticsFiltering` function. I have some log statements as you can see, to let me know if my `TelemetryInitializer` is being added correctly (I need to pre-process the data sent to Application Insights).

However, the `console.log("Adding to context");` is never reached. So I debugged the application and inspected `this.appInsightsService` and it turns out the `context` is set to undefined.

Why is the `context` set to undefined and how do I get it to initialize correctly? None of my `TelementryInitializer`s are working because of this.

**Comment by atbe-crowe** [comment]
To add, the `appInsights.context` **is** initialized, so I don't see why the `AppInsightsService` doesn't have the correct context

**Comment by graemeenglish** [comment]
I had the same issue although I couldn't use the library directly since I'm using ionic and Angular 5. Solution was to change the following line:
this.context = AppInsights.context
to
this.queue.push(() => (this.context = AppInsights.context));

**Comment by MarkPieszak** [comment]
For it to work you'll need to have it at your Root NgModule, otherwise create a shared NgModule with the ApplicationInsightsModule (with no .forRoot) that you use within your child modules, _or_ grab the AppInsightsService via the `Injector`.

```ts
constructor(
  private injector: Injector
) {
  this.injector.get(AppInsightsService);
}
```

Closing out for now but let me know if you're still having trouble here!

**Comment by FraserKillip** [comment]
I have put in a fix for this with #57 since I ran into the same problem. My problem was that AppInsights was not set up when it tried to set a reference to the context.

**Comment by MarkPieszak** [comment]
@atbe-crowe @graemeenglish 
Fixed via #57 - released as version 6.0.0
